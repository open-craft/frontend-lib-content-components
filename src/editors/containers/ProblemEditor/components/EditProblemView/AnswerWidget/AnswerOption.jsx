import React, { memo, useEffect, useState } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import {
  Col, Collapsible, Icon, IconButton, Form, Row,
} from '@edx/paragon';
import { AddComment, Delete } from '@edx/paragon/icons';
import { FormattedMessage, injectIntl, intlShape } from '@edx/frontend-platform/i18n';

import messages from './messages';
import { actions, selectors } from '../../../../../data/redux';
import { answerOptionProps } from '../../../../../data/services/cms/types';
import { ProblemTypeKeys } from '../../../../../data/constants/problem';

export const AnswerOption = ({
  answer,
  hasSingleAnswer,
  // injected
  intl,
  // redux
  deleteAnswer,
  updateAnswer,
  problemType,
}) => {
  const [isFeedbackVisible, setIsFeedbackVisible] = useState(false);
  const removeAnswer = () => deleteAnswer({ id: answer.id });
  const setAnswer = (payload) => updateAnswer({ id: answer.id, hasSingleAnswer, ...payload });

  useEffect(() => {
    // Show feedback fields if feedback is present
    setIsFeedbackVisible(isVisible => (
      !!answer.selectedFeedback || !!answer.unselectedFeedback || !!answer.feedback || isVisible
    ));
  }, [answer]);

  const toggleFeedback = (open) => {
    // Do not allow to hide if feedback is added
    if (!!answer.selectedFeedback || !!answer.unselectedFeedback || !!answer.feedback) {
      setIsFeedbackVisible(true);
      return;
    }
    setIsFeedbackVisible(open);
  };

  const Checker = () => {
    let CheckerType = Form.Checkbox;
    if (hasSingleAnswer) {
      CheckerType = Form.Radio;
    }
    return (
      <CheckerType
        className="pl-4 mt-3"
        value={answer.id}
        onChange={(e) => setAnswer({ correct: e.target.checked })}
        defaultChecked={answer.correct}
      >
        {answer.id}
      </CheckerType>
    );
  };

  const feedbackControl = ({
    feedback, onChange, labelMessage, labelMessageBoldUnderline, key,
  }) => (
    <Form.Group key={key}>
      <Form.Label className="mb-3">
        <FormattedMessage
          {...labelMessage}
          values={{
            answerId: answer.id,
            boldunderline: <b><u><FormattedMessage {...labelMessageBoldUnderline} /></u></b>,
          }}
        />
      </Form.Label>
      <Form.Control
        placeholder={intl.formatMessage(messages.feedbackPlaceholder)}
        value={feedback}
        onChange={onChange}
      />
    </Form.Group>
  );

  const displayFeedbackControl = () => {
    if (problemType !== ProblemTypeKeys.MULTISELECT) {
      return feedbackControl({
        key: `feedback-${answer.id}`,
        feedback: answer.feedback,
        onChange: (e) => setAnswer({ feedback: e.target.value }),
        labelMessage: messages.selectedFeedbackLabel,
        labelMessageBoldUnderline: messages.selectedFeedbackLabelBoldUnderlineText,
      });
    }
    return [
      feedbackControl({
        key: `selectedfeedback-${answer.id}`,
        feedback: answer.selectedFeedback,
        onChange: (e) => setAnswer({ selectedFeedback: e.target.value }),
        labelMessage: messages.selectedFeedbackLabel,
        labelMessageBoldUnderline: messages.selectedFeedbackLabelBoldUnderlineText,
      }),
      feedbackControl({
        key: `unselectedfeedback-${answer.id}`,
        feedback: answer.unselectedFeedback,
        onChange: (e) => setAnswer({ unselectedFeedback: e.target.value }),
        labelMessage: messages.unSelectedFeedbackLabel,
        labelMessageBoldUnderline: messages.unSelectedFeedbackLabelBoldUnderlineText,
      }),
    ];
  };

  return (
    <Collapsible.Advanced
      open={isFeedbackVisible}
      onToggle={toggleFeedback}
      className="collapsible-card"
    >
      <Row className="my-2">

        <Col xs={1}>
          <Checker />
        </Col>

        <Col xs={10}>
          <Form.Control
            as="textarea"
            rows={1}
            value={answer.title}
            onChange={(e) => setAnswer({ title: e.target.value })}
            placeholder={intl.formatMessage(messages.answerTextboxPlaceholder)}
          />

          <Collapsible.Body>
            <div className="bg-dark-100 p-4 mt-3">
              {displayFeedbackControl()}
            </div>
          </Collapsible.Body>
        </Col>

        <Col xs={1} className="d-inline-flex mt-1">
          <Collapsible.Trigger>
            <IconButton
              src={AddComment}
              iconAs={Icon}
              alt={intl.formatMessage(messages.feedbackToggleIconAltText)}
              variant="primary"
            />
          </Collapsible.Trigger>
          <IconButton
            src={Delete}
            iconAs={Icon}
            alt={intl.formatMessage(messages.answerDeleteIconAltText)}
            onClick={removeAnswer}
            variant="primary"
          />
        </Col>

      </Row>
    </Collapsible.Advanced>
  );
};

AnswerOption.propTypes = {
  answer: answerOptionProps.isRequired,
  hasSingleAnswer: PropTypes.bool.isRequired,
  // injected
  intl: intlShape.isRequired,
  // redux
  deleteAnswer: PropTypes.func.isRequired,
  updateAnswer: PropTypes.func.isRequired,
  problemType: PropTypes.string.isRequired,
};

export const mapStateToProps = (state) => ({
  problemType: selectors.problem.problemType(state),
});

export const mapDispatchToProps = {
  deleteAnswer: actions.problem.deleteAnswer,
  updateAnswer: actions.problem.updateAnswer,
};
export default injectIntl(connect(mapStateToProps, mapDispatchToProps)(memo(AnswerOption)));
