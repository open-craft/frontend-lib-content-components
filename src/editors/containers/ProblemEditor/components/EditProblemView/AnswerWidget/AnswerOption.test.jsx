import React from 'react';
import { shallow } from 'enzyme';
import { formatMessage } from '../../../../../../testUtils';
import { AnswerOption } from './AnswerOption';

describe('AnswerOption', () => {
  const answerWithOnlyFeedback = {
    id: 'A',
    title: 'Answer 1',
    correct: true,
    feedback: 'some feedback',
  };
  const answerWithSelectedUnselectedFeedback = {
    id: 'A',
    title: 'Answer 1',
    correct: true,
    selectedFeedback: 'selected feedback',
    unselectedFeedback: 'unselected feedback',
  };

  const props = {
    hasSingleAnswer: false,
    answer: answerWithOnlyFeedback,
    // inject
    intl: { formatMessage },
    deleteAnswer: jest.fn(),
    updateAnswer: jest.fn(),
  };
  describe('render', () => {
    test('snapshot: renders correct option with feedback', () => {
      expect(shallow(<AnswerOption {...props} />)).toMatchSnapshot();
    });
    test('snapshot: renders correct option with selected unselected feedback', () => {
      expect(shallow(<AnswerOption {...props} answer={answerWithSelectedUnselectedFeedback} />)).toMatchSnapshot();
    });
  });
});
