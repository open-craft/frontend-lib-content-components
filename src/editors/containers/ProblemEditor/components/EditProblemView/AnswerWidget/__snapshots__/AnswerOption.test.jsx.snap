// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnswerOption render snapshot: renders correct option with feedback 1`] = `
<Collapsible.Advanced
  className="collapsible-card"
  onToggle={[Function]}
  open={false}
>
  <Row
    className="my-2"
  >
    <Col
      xs={1}
    >
      <Checker />
    </Col>
    <Col
      xs={10}
    >
      <Form.Control
        as="textarea"
        onChange={[Function]}
        placeholder="Enter an answer"
        rows={1}
        value="Answer 1"
      />
      <Collapsible.Body>
        <div
          className="bg-dark-100 p-4 mt-3"
        >
          <Form.Group
            key="feedback-A"
          >
            <Form.Label
              className="mb-3"
            >
              <FormattedMessage
                defaultMessage="Show following feedback when {answerId} {boldunderline}:"
                description="Label text for feedback if option is selected"
                id="authoring.answerwidget.feedback.selected.label"
                values={
                  Object {
                    "answerId": "A",
                    "boldunderline": <b>
                      <u>
                        <FormattedMessage
                          defaultMessage="is selected"
                          description="Bold & underlined text for feedback if option is selected"
                          id="authoring.answerwidget.feedback.selected.label.boldunderline"
                        />
                      </u>
                    </b>,
                  }
                }
              />
            </Form.Label>
            <Form.Control
              onChange={[Function]}
              placeholder="Feedback message"
              value="some feedback"
            />
          </Form.Group>
        </div>
      </Collapsible.Body>
    </Col>
    <Col
      className="d-inline-flex mt-1"
      xs={1}
    >
      <Collapsible.Trigger>
        <IconButton
          alt="Toggle feedback"
          iconAs="Icon"
          variant="primary"
        />
      </Collapsible.Trigger>
      <IconButton
        alt="Delete answer"
        iconAs="Icon"
        onClick={[Function]}
        variant="primary"
      />
    </Col>
  </Row>
</Collapsible.Advanced>
`;

exports[`AnswerOption render snapshot: renders correct option with selected unselected feedback 1`] = `
<Collapsible.Advanced
  className="collapsible-card"
  onToggle={[Function]}
  open={false}
>
  <Row
    className="my-2"
  >
    <Col
      xs={1}
    >
      <Checker />
    </Col>
    <Col
      xs={10}
    >
      <Form.Control
        as="textarea"
        onChange={[Function]}
        placeholder="Enter an answer"
        rows={1}
        value="Answer 1"
      />
      <Collapsible.Body>
        <div
          className="bg-dark-100 p-4 mt-3"
        >
          <Form.Group
            key="selectedfeedback-A"
          >
            <Form.Label
              className="mb-3"
            >
              <FormattedMessage
                defaultMessage="Show following feedback when {answerId} {boldunderline}:"
                description="Label text for feedback if option is selected"
                id="authoring.answerwidget.feedback.selected.label"
                values={
                  Object {
                    "answerId": "A",
                    "boldunderline": <b>
                      <u>
                        <FormattedMessage
                          defaultMessage="is selected"
                          description="Bold & underlined text for feedback if option is selected"
                          id="authoring.answerwidget.feedback.selected.label.boldunderline"
                        />
                      </u>
                    </b>,
                  }
                }
              />
            </Form.Label>
            <Form.Control
              onChange={[Function]}
              placeholder="Feedback message"
              value="selected feedback"
            />
          </Form.Group>
          <Form.Group
            key="unselectedfeedback-A"
          >
            <Form.Label
              className="mb-3"
            >
              <FormattedMessage
                defaultMessage="Show following feedback when {answerId} {boldunderline}:"
                description="Label text for feedback if option is not selected"
                id="authoring.answerwidget.feedback.unselected.label"
                values={
                  Object {
                    "answerId": "A",
                    "boldunderline": <b>
                      <u>
                        <FormattedMessage
                          defaultMessage="is not selected"
                          description="Bold & underlined text for feedback if option is not selected"
                          id="authoring.answerwidget.feedback.unselected.label.boldunderline"
                        />
                      </u>
                    </b>,
                  }
                }
              />
            </Form.Label>
            <Form.Control
              onChange={[Function]}
              placeholder="Feedback message"
              value="unselected feedback"
            />
          </Form.Group>
        </div>
      </Collapsible.Body>
    </Col>
    <Col
      className="d-inline-flex mt-1"
      xs={1}
    >
      <Collapsible.Trigger>
        <IconButton
          alt="Toggle feedback"
          iconAs="Icon"
          variant="primary"
        />
      </Collapsible.Trigger>
      <IconButton
        alt="Delete answer"
        iconAs="Icon"
        onClick={[Function]}
        variant="primary"
      />
    </Col>
  </Row>
</Collapsible.Advanced>
`;
